web_src/js/markdown/checkboxes.js

```diff
@@ -0,0 +1,70 @@
+const checkboxMarkdownPattern = /\[[ x]]/g;
+
+/**
+ * Attaches `change` handlers to markdown rendered checkboxes in comments.
+ * When a checkbox value changes, the corresponding [ ] or [x] in the markdown string is set accordingly and sent to the server.
+ * On success it updates the raw-content on error it resets the checkbox to its original value.
+ */
+export default function initMarkdownCheckboxes() {
+  $('.comment .segment').each((_, segment) => {
+    const $segment = $(segment);
+    const $checkboxes = $segment.find('.render-content.markdown input:checkbox');
+
+    const onChange = async (ev, cbIndex) => {
+      const $cb = $(ev.target);
+      const checkboxMarkdown = $cb.is(':checked') ? '[x]' : '[ ]';
+
+      const $rawContent = $segment.find('.raw-content');
+      const oldContent = $rawContent.text();
+      const newContent = oldContent.replace(checkboxMarkdownPattern, replaceNthMatchWith(cbIndex, checkboxMarkdown));
```

Comment:
how does this work on:

    -[ ]: I'm a checkbox
    
    ```
    -[ ]: how you write a checkbox
    ```
    
    -[ ]: I'm another checkbox

?

Reply:
:point_up: this will set an x in the code-block if you click on the last checkbox :/

Reply:
@j2L4e nice pull, it just need to ignore things in code-blocks

Reply:
You're right. The same applies to `inline code` and that's complicating things. I'm not sure if RegEx is going to cut it.

Does attaching a `data-raw-position-start` prop to the rendered checkboxes seem feasible to you? Somewhere around here? I'd expect the AST node to hold information about its origin (i.e. position in raw text)
https://github.com/go-gitea/gitea/blob/master/modules/markup/markdown/goldmark.go#L359

If all checkboxes had that info attached this would be trivial and avoid brittle client-side parsing shenanigans.

I'm unfamiliar with goldmark and go in general, so if one of you guys could do it that would be awesome.

Reply:
Lines()[0].Start might provide the info - but you may have to crawl the children to find the first Text child with a Segment.

Reply:
@j2L4e :point_up:

Reply:
the below diff would add a data-index attribute to the checkbox that should provide the index in the raw markdown of the checkbox itself.

<details><summary> attached diff </summary>

```diff
diff --git a/modules/markup/markdown/goldmark.go b/modules/markup/markdown/goldmark.go
index f9fd6eb97..e5ff04539 100644
--- a/modules/markup/markdown/goldmark.go
+++ b/modules/markup/markdown/goldmark.go
@@ -8,8 +8,10 @@ import (
 	"bytes"
 	"fmt"
 	"regexp"
+	"strconv"
 	"strings"
 
+	"code.gitea.io/gitea/modules/log"
 	"code.gitea.io/gitea/modules/markup"
 	"code.gitea.io/gitea/modules/markup/common"
 	"code.gitea.io/gitea/modules/setting"
@@ -355,15 +357,38 @@ func (r *HTMLRenderer) renderTaskCheckBoxListItem(w util.BufWriter, source []byt
 		if r.XHTML {
 			end = " />"
 		}
+
+		idx := -1
+		if err := ast.Walk(n, func(n ast.Node, entering bool) (ast.WalkStatus, error) {
+			textNode, ok := n.(*ast.Text)
+			if ok {
+				idx = textNode.Segment.Start
+				return ast.WalkStop, nil
+			}
+
+			rawHTMLNode, ok := n.(*ast.RawHTML)
+			if ok && rawHTMLNode.Segments != nil && rawHTMLNode.Segments.Len() > 0 {
+				idx = rawHTMLNode.Segments.At(0).Start
+				return ast.WalkStop, nil
+			}
+
+			return ast.WalkContinue, nil
+		}); err != nil {
+			return ast.WalkStop, err
+		}
+
+		log.Info("here %d", idx)
+
 		var err error
 		if n.IsChecked {
-			_, err = w.WriteString(`<input type="checkbox" disabled="" checked=""` + end)
+			_, err = w.WriteString(`<input type="checkbox" disabled="" checked="" data-index="` + strconv.Itoa(idx) + `"` + end)
 		} else {
-			_, err = w.WriteString(`<input type="checkbox" disabled=""` + end)
+			_, err = w.WriteString(`<input type="checkbox" disabled="" data-index="` + strconv.Itoa(idx) + `"` + end)
 		}
 		if err != nil {
 			return ast.WalkStop, err
 		}
+
 		fc := n.FirstChild()
 		if fc != nil {
 			if _, ok := fc.(*ast.TextBlock); !ok {
diff --git a/modules/markup/sanitizer.go b/modules/markup/sanitizer.go
index 9214a75fb..3106437ba 100644
--- a/modules/markup/sanitizer.go
+++ b/modules/markup/sanitizer.go
@@ -43,7 +43,7 @@ func ReplaceSanitizer() {
 
 	// Checkboxes
 	sanitizer.policy.AllowAttrs("type").Matching(regexp.MustCompile(`^checkbox$`)).OnElements("input")
-	sanitizer.policy.AllowAttrs("checked", "disabled").OnElements("input")
+	sanitizer.policy.AllowAttrs("checked", "disabled", "data-index").OnElements("input")
 
 	// Custom URL-Schemes
 	sanitizer.policy.AllowURLSchemes(setting.Markdown.CustomURLSchemes...)
```

</details>

Reply:
Great, thank you!
I'll update this once I'm ready